### DISCUSSION TITLE
Re-define `lychee` as both a link checker and a link linter

### DISCUSSION THREAD

**Discussion by lebensterben** [lycheeverse/lychee#880] [Ideas]
Currently, `lychee` is just a link ___checker___, which conceptually has the following components:
1. It parses a config file or CLI argument, and turns it into an internal configuration
2. An __overseer__  is created given the configurations. It serves several purposes:
   - Detection: It detects/scrape HTML links and EMAIL addresses from the input file(s)
   - Filtering: It checks whether the link should be checked or skipped
   - Dispatcher: It sends URLs to __workers__ who verify whether a given URL is valid by checking whether it's accessible.
   - Logging: It keeps an record of URL it encounters and the status of each URL. (Valid, Ignored, Timeout, etc)
3. The result is returned either as CLI output or to a log file.

---

We can extend `lychee` so that it's also a link ___linter___.

Note that the __overseer__ hands the job to its __workers__, and __workers__ are only responsible for validating the URLs. The __linting__ facility should also be done by __overseer__.

Conceptually, __linting__ could happen before and/or after the __overseer__ dispatches the job. For example:
- Before: If the user wants to deny a certain pattern in URL, the __overseer__ don't need to send URLs that match the pattern to the __workers__, instead it directly logs it.
- After: If the user wants to avoid the use of absolute/relative links, the __overseer__ first verify whether a link is valid, and if it is, then it check whether it's in the desired style before logging.
- Around (before and after): If the user wants to use HTTPS links whenever possible, the __overseer__ first sends an HTTP link to __workers__, and if it's valid the __overseer__ resends it HTTPS counterpart the __workers__, and logs the result accordingly.

**Comment by untitaker** [comment]
This is labelled as `design-feedback` because it already goes into the nitty-gritty of which code unit does what, but from what i can tell what is actually proposed here is to expand lychee's feature scope such that it can be used to enforce policies on HTML, such as the already mentioned "need to use HTTPs links everywhere" and "links can't match this pattern".

You can already enforce some basic pattern-policies today: Use `--dump` and a wrapper shell script. What can't be done is the HTTPs enforcement in the way you imagine it.

Is the list of policies in your OP exhaustive?

**Comment by mre** [comment]
@lebensterben any thoughts?

**Comment by untitaker** [comment]
prior art: https://github.com/wjdp/htmltest

**Comment by mre** [comment]
Link validation is a whole other use-case with a lot of design decisions to consider along the way. We have to be careful to keep the scope manageable. I guess we can commit to the following:
* enforce HTTPS links: This was added in #192 
* allow/block list, see https://github.com/lycheeverse/lychee/issues/486. There is a workaround using `--dump`.

Outside of that, I'd probably defer to other tools (e.g. htmltest that @untitaker mentioned) or workarounds using `--dump` for now.

**Comment by untitaker** [comment]
I checked OP again, what you currently cannot do is hook into before/after link traversal for linting, or define your own link extraction logic. `--dump` is insufficient there.

eg you may want to lint a link and based on the linting decide whether to follow the link.

I wonder if OP meant to build a scripting platform on top of lychee where the user could hook custom logic into any of those stages, and that's why internals are discussed so much in detail.

**Comment by mre** [comment]
FYI: I've converted this from and issue into a discussion as we still need to discuss more details before taking any action.

**Comment by dkarlovi** [comment]
Additional linting opportunity 
https://github.com/lycheeverse/lychee/pull/1777#issuecomment-3144386405
